src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -558,67 +638,79 @@ std::pair<std::vector<String>, bool> ReplicatedMergeTreeSinkImpl<async_insert>::
     ///
     /// metadata_snapshot->check(part->getColumns());
 
-    const String temporary_part_relative_path = part->getDataPartStorage().getPartDirectory();
+    auto block_id_path = getBlockIdPath(storage.zookeeper_path, block_id);
 
-    /// There is one case when we need to retry transaction in a loop.
-    /// But don't do it too many times - just as defensive measure.
-    size_t loop_counter = 0;
-    constexpr size_t max_iterations = 10;
+    CommitRetryContext retry_context;
 
-    bool is_already_existing_part = false;
+    ZooKeeperRetriesControl retries_ctl("commitPart", zookeeper_retries_info, context->getProcessListElement());
 
-    /// for retries due to keeper error
-    bool part_committed_locally_but_zookeeper = false;
-    bool part_was_deduplicated = false;
-    Coordination::Error write_part_info_keeper_error = Coordination::Error::ZOK;
-    std::vector<String> conflict_block_ids;
+    auto resolve_duplicate_stage = [&] () -> CommitRetryContext::Stages
+    {
+        if constexpr (async_insert)
+        {
+            throw Exception(ErrorCodes::LOGICAL_ERROR,
+                            "Conflict block ids and block number lock should not "
+                            "be empty at the same time for async inserts");
+        }
+        else
+        {
+            /// This block was already written to some replica. Get the part name for it.
+            /// Note: race condition with DROP PARTITION operation is possible. User will get "No node" exception and it is Ok.
+            retry_context.actual_part_name = zookeeper->get(block_id_path);
 
-    ZooKeeperRetriesControl retries_ctl("commitPart", zookeeper_retries_info, context->getProcessListElement());
-    retries_ctl.retryLoop([&]()
+            bool exists_locally = bool(storage.getActiveContainingPart(retry_context.actual_part_name));
+
+            if (exists_locally)
+                ProfileEvents::increment(ProfileEvents::DuplicatedInsertedBlocks);
+
+            LOG_INFO(log, "Block with ID {} {} as part {}; ignoring it.",
+                     block_id,
+                     exists_locally ? "already exists locally" : "already exists on other replicas",
+                     retry_context.actual_part_name);
+
+            retry_context.part_was_deduplicated = true;
+
+            return CommitRetryContext::SUCCESS;
+        }
+    };
+
+    auto resolve_uncertain_commit_stage = [&] () -> CommitRetryContext::Stages
     {
-        zookeeper->setKeeper(storage.getZooKeeper());
-        if (storage.is_readonly)
+        /// check that info about the part was actually written in zk
```

Comment:
In case it helps visualizing, here is how I intented to introduce it with the old code: https://github.com/ClickHouse/ClickHouse/pull/56917/files#diff-d34dfcef2ee825d0184c61973c0c3c8606d68975870f22c9ddd1501f7ff88e42R926-R969

Basically I created a new ZooKeeperRetriesControl just to check the status of the multi operation, with 3 options:
* We contact ZK and verify that the write worked -> Do whatever we do on success.
* We contact ZK and verify that the write did not work -> Restart the process. Following https://github.com/ClickHouse/ClickHouse/pull/57205/files#r1411190204, it means acquiring a new block number and all operations based on it (basically starting from scratch). This must be clean (i.e. not leave nodes in ZK, which should be ok now).
* We are unable to contact ZK again so we don't know what happened. No need to keep retrying (it's been retried in this loop), commit the transaction, add the part to the check list and throw UNEXPECTED_ZOOKEEPER_ERROR.

That is, the error handling happens inside the `commit_new_part_stage` step and the `resolve_uncertain_commit_stage` would disappear and IMO simplifies the logic, as before it was done at the beginning of the loop and that's really confusing when you see the code for the first time.